rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -109,34 +109,54 @@ void write_vertex_array_data_to_buffer(void *buffer, u32 first, u32 count, size_
 	}
 }
 
-template<typename IndexType, typename DstType, typename SrcType>
-void expandIndexedTriangleFan(DstType *dst, const SrcType *src, size_t indexCount)
+template <typename T>
+struct VMRead
+{
+	static T exec(u32 address);
+};
+
+template<>
+struct VMRead<u16>
+{
+	static u16 exec(u32 address)
+	{
+		return vm::ps3::read16(address);
+	}
+};
+
+template<>
+struct VMRead<u32>
+{
+	static u32 exec(u32 address)
+	{
+		return vm::ps3::read32(address);
+	}
+};
+
+template<typename IndexType>
+void expandIndexedTriangleFan(char *dst, u32 address, size_t indexCount)
 {
-	IndexType *typedDst = reinterpret_cast<IndexType *>(dst);
-	const IndexType *typedSrc = reinterpret_cast<const IndexType *>(src);
 	for (unsigned i = 0; i < indexCount - 2; i++)
 	{
-		typedDst[3 * i] = typedSrc[0];
-		typedDst[3 * i + 1] = typedSrc[i + 2 - 1];
-		typedDst[3 * i + 2] = typedSrc[i + 2];
+		(IndexType&)dst[(3 * i) * sizeof(IndexType)] = VMRead<IndexType>::exec(address);
```

Comment:
This is a bit... too much. `vm::ps3::_ref<IndexType>(address)` must be sufficient if you didn't mean something extraordinary in the VMRead template. `vm::_ref`, if it's RSX-specific file and ps3 namespace can be correctly used.